search: Check if user is subscribed before requesting stream data.

laurynmm opened the pull request:
Normally a user will not have any stream data for private streams that they are not subscribed to, but admin users do have access to basic information about these private streams for admin purposes.

When building search suggestions, instead of assuming that a user with the basic stream information has access to the message data for the stream, check first to see if the user is subscribed before sending the request to the server for more stream topic data.

---

**Screenshots and screen captures:**

<details>
<summary>Pre-changes: 400 stack trace in dev environment</summary>

![Screenshot from 2022-10-19 19-41-35](https://user-images.githubusercontent.com/63245456/196765709-f57ad72d-6ca1-4709-ab9d-bb32b3a14e47.png)
</details>

---

**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.

timabbott commented:
Good find! I think we want a slightly different check; the request will succeed if `access_stream_by_id(user_profile, stream_id)` will succeed, which includes public streams (if you're not a guest). 

I'm not sure which function is the right check; maybe  `stream_data.can_preview` or `stream_data.can_toggle_subscription()` is correct?

laurynmm commented:
> Good find! I think we want a slightly different check; the request will succeed if `access_stream_by_id(user_profile, stream_id)` will succeed, which includes public streams (if you're not a guest).
> 
> I'm not sure which function is the right check; maybe `stream_data.can_toggle_subscription()` is the right check?

Yeah, after looking at `check_basic_stream_access` in `zerver/lib/streams.py`, it seems like we want a combination of a is web public stream check and the `stream_data.can_toggle_subscription()` check. Updated the PR for those two checks.

The backend also checks for admin permissions, but that's to access the stream name, description and subscribers. In this situation, we're trying to get stream topics, which the admin shouldn't have access to in a private stream unless they are subscribed. These changes also filter out similar 400 responses for a user that has unsubscribed from a private stream.

I removed the comment from the original post as I wasn't convinced we need one here. I also noticed we call `stream_topic_history_util.get_server_history(stream_id)` in a similar way in `composebox_typeahead.js`. Seems like a much less likely place one might see this happen.

timabbott commented:
This is great; merged after extracting a `can_access_topic_history` helper that we might be able to reuse somewhere.

Thanks @laurynmm!
